src/Processors/QueryPlan/Optimizations/removeRedundantOrderBy.cpp

```diff
@@ -0,0 +1,190 @@
+#include <AggregateFunctions/AggregateFunctionFactory.h>
+#include <Processors/QueryPlan/AggregatingStep.h>
+#include <Processors/QueryPlan/ExpressionStep.h>
+#include <Processors/QueryPlan/FillingStep.h>
+#include <Processors/QueryPlan/ITransformingStep.h>
+#include <Processors/QueryPlan/LimitByStep.h>
+#include <Processors/QueryPlan/LimitStep.h>
+#include <Processors/QueryPlan/Optimizations/Optimizations.h>
+#include <Processors/QueryPlan/ReadFromMergeTree.h>
+#include <Processors/QueryPlan/SortingStep.h>
+#include <Common/logger_useful.h>
+#include <Common/typeid_cast.h>
+
+namespace DB::QueryPlanOptimizations
+{
+const char * stepName(const QueryPlan::Node * node)
+{
+    IQueryPlanStep * current_step = node->step.get();
+    return typeid(*current_step).name();
+}
+
+void printStepName(const char * prefix, const QueryPlan::Node * node)
+{
+    LOG_DEBUG(&Poco::Logger::get("RedundantOrderBy"), "{}: {}: {}", prefix, stepName(node), reinterpret_cast<void *>(node->step.get()));
+}
+
+struct FrameWithParent
+{
+    QueryPlan::Node * node = nullptr;
+    QueryPlan::Node * parent_node = nullptr;
+    size_t next_child = 0;
+};
+
+using StackWithParent = std::vector<FrameWithParent>;
+
+bool checkForStatefulFunctions(const StackWithParent & stack, const QueryPlan::Node * node_affect_order)
+{
+    chassert(!stack.empty());
+    chassert(typeid_cast<const SortingStep *>(stack.back().node->step.get()));
+
+    /// skip element on top of stack since it's sorting
+    for (StackWithParent::const_reverse_iterator it = stack.rbegin() + 1; it != stack.rend(); ++it)
+    {
+        const auto * node = it->node;
+        if (node == node_affect_order)
+            break;
+
+        const auto * step = node->step.get();
+
+        const auto * expr = typeid_cast<const ExpressionStep *>(step);
+        if (expr)
+        {
+            if (expr->getExpression()->hasStatefulFunctions())
+                return true;
+        }
+        else
+        {
+            const auto * trans = typeid_cast<const ITransformingStep *>(step);
+            if (!trans)
+                break;
+
+            if (!trans->getDataStreamTraits().preserves_sorting)
+                break;
+        }
+    }
+    return false;
+}
+
+void tryRemoveRedundantOrderBy(QueryPlan::Node * root)
```

Comment:
I guess, it'd be nicer to implement it as a class like visitor with corresponding methods for top-down/bottom-up traversing